flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStream.java

```diff
@@ -163,14 +251,34 @@
 			null,
 			false);
 
+		final OutputTag<L> outputTag = new OutputTag<L>("dummy-timeouted", leftTypeInfo);
+
+		final SingleOutputStreamOperator<R> mainStream = CEPOperatorUtils.createTimeoutPatternStream(
+			inputStream,
+			pattern,
+			clean(patternSelectFunction),
+			rightTypeInfo,
+			outputTag,
+			clean(patternTimeoutFunction));
+
+		final DataStream<L> timeoutedStream = mainStream.getSideOutput(outputTag);
+
 		TypeInformation<Either<L, R>> outTypeInfo = new EitherTypeInfo<>(leftTypeInfo, rightTypeInfo);
 
-		return patternStream.map(
-			new PatternSelectTimeoutMapper<>(
-				patternStream.getExecutionEnvironment().clean(patternSelectFunction),
-				patternStream.getExecutionEnvironment().clean(patternTimeoutFunction)
-			)
-		).returns(outTypeInfo);
+		return mainStream.connect(timeoutedStream).map(new CoMapFunction<R, L, Either<L, R>>() {
```

Comment:
I think using an anonymous inner class here can be problematic because it will capture the `PatternStream` in it's closure, if I'm not mistaken.